File: r_app/simulate.R
# =====================================================================
# Simulate synthetic A/B-test events directly into the SQLite event log.
#
# Useful for developing / demoing the analysis pipeline before enough
# real users have interacted with the deployed Shiny app.
#
# Usage
# -----
#   Rscript simulate.R --n 2000 --reset
#   Rscript simulate.R --n 5000 --seed 7
# =====================================================================

suppressPackageStartupMessages({
  library(DBI)
  library(RSQLite)
})

SCHEMA <- "
CREATE TABLE IF NOT EXISTS events (
    id              INTEGER PRIMARY KEY AUTOINCREMENT,
    user_id         TEXT    NOT NULL,
    experiment      TEXT    NOT NULL,
    variant         TEXT    NOT NULL,
    event_type      TEXT    NOT NULL,
    created_at      TEXT    NOT NULL,
    user_agent      TEXT,
    referrer        TEXT,
    payload         TEXT
);
CREATE INDEX IF NOT EXISTS idx_events_experiment_variant
    ON events (experiment, variant, event_type);
"

parse_args <- function(args) {
  out <- list(
    db = Sys.getenv("AB_DB_PATH", unset = "ab_test.db"),
    n = 2000L, seed = 42L, reset = FALSE,
    discount_a = 0.10, discount_b = 0.15,
    form_a = 0.50, form_b = 0.35
  )
  i <- 1
  while (i <= length(args)) {
    a <- args[[i]]
    if (a == "--db")         { out$db    <- args[[i + 1]];                i <- i + 2; next }
    if (a %in% c("--n"))     { out$n     <- as.integer(args[[i + 1]]);    i <- i + 2; next }
    if (a %in% c("--seed"))  { out$seed  <- as.integer(args[[i + 1]]);    i <- i + 2; next }
    if (a == "--reset")      { out$reset <- TRUE;                          i <- i + 1; next }
    if (a == "--discount-a") { out$discount_a <- as.numeric(args[[i + 1]]); i <- i + 2; next }
    if (a == "--discount-b") { out$discount_b <- as.numeric(args[[i + 1]]); i <- i + 2; next }
    if (a == "--form-a")     { out$form_a     <- as.numeric(args[[i + 1]]); i <- i + 2; next }
    if (a == "--form-b")     { out$form_b     <- as.numeric(args[[i + 1]]); i <- i + 2; next }
    stop("Unknown argument: ", a)
  }
  out
}

ensure_schema <- function(con) {
  for (stmt in strsplit(SCHEMA, ";")[[1]]) {
    stmt <- trimws(stmt)
    if (nzchar(stmt)) dbExecute(con, stmt)
  }
}

random_uid <- function(n) {
  replicate(n, paste(sample(c(letters, 0:9), 24, replace = TRUE), collapse = ""))
}

simulate <- function(opts) {
  set.seed(opts$seed)
  dir.create(dirname(opts$db), showWarnings = FALSE, recursive = TRUE)
  con <- dbConnect(SQLite(), opts$db)
  on.exit(dbDisconnect(con), add = TRUE)
  ensure_schema(con)
  if (opts$reset) dbExecute(con, "DELETE FROM events")

  uids <- random_uid(opts$n)
  experiments <- sample(c("discount", "form_length"), opts$n, replace = TRUE)
  variants    <- sample(c("A", "B"), opts$n, replace = TRUE)

  p <- ifelse(experiments == "discount",
              ifelse(variants == "A", opts$discount_a, opts$discount_b),
              ifelse(variants == "A", opts$form_a,     opts$form_b))

  now <- Sys.time()
  rows <- list()
  for (i in seq_len(opts$n)) {
    ts_imp <- format(now - (opts$n - i), "%Y-%m-%dT%H:%M:%S", tz = "UTC")
    rows[[length(rows) + 1L]] <- list(
      uids[i], experiments[i], variants[i], "impression",
      ts_imp, "Simulator/1.0", "", NA_character_
    )
    if (runif(1) < p[i]) {
      ts_c <- format(now - (opts$n - i) + runif(1, 5, 120),
                     "%Y-%m-%dT%H:%M:%S", tz = "UTC")
      rows[[length(rows) + 1L]] <- list(
        uids[i], experiments[i], variants[i], "conversion",
        ts_c, "Simulator/1.0", "", "simulated=1"
      )
    }
  }
  # Build a data.frame from the rows in a single shot, then append it.
  df <- do.call(
    rbind.data.frame,
    lapply(rows, function(r) data.frame(
      user_id     = r[[1]],
      experiment  = r[[2]],
      variant     = r[[3]],
      event_type  = r[[4]],
      created_at  = r[[5]],
      user_agent  = r[[6]],
      referrer    = r[[7]],
      payload     = r[[8]],
      stringsAsFactors = FALSE
    ))
  )
  dbAppendTable(con, "events", df)

  total <- dbGetQuery(con, "SELECT COUNT(*) AS n FROM events")$n
  cat(sprintf(
    "Simulated %d users -> wrote %d events to %s (total events now: %d).\n",
    opts$n, nrow(df), opts$db, total))
}

if (!interactive()) {
  simulate(parse_args(commandArgs(trailingOnly = TRUE)))
}

File: simulate.py
"""
Simulate synthetic A/B-test data directly in the SQLite event log.

Useful for development, demo, and to sanity-check the analysis pipeline
when few real users have interacted with the app yet.

Each simulated user sees **one** experiment (discount OR form_length),
is randomly assigned to A or B, always records an impression, and then
converts with a pre-configured probability.  The resulting database is
fully compatible with ``analysis.py`` and the ``/stats`` dashboard.

Example
-------
    python simulate.py --n 2000 --seed 7
    python simulate.py --reset --n 5000

By default, the "true" conversion rates are:

    discount     A = 0.10   B = 0.15       (B wins, +5 pp)
    form_length  A = 0.50   B = 0.35       (A wins, +15 pp)

so that the analysis script can demonstrate both a winning and losing
variant.  Override them with --discount-a / --discount-b / --form-a /
--form-b.
"""

from __future__ import annotations

import argparse
import os
import random
import sqlite3
import uuid
from contextlib import closing
from datetime import datetime, timedelta, timezone
from pathlib import Path

from flask_app import DB_PATH as DEFAULT_DB_PATH, SCHEMA


def ensure_schema(conn: sqlite3.Connection) -> None:
    conn.executescript(SCHEMA)
    conn.commit()


def reset_events(conn: sqlite3.Connection) -> None:
    conn.execute("DELETE FROM events")
    conn.commit()


def simulate(
    db_path: Path,
    n_users: int,
    *,
    discount_rates: tuple[float, float],
    form_rates: tuple[float, float],
    seed: int = 0,
    reset: bool = False,
) -> None:
    random.seed(seed)

    db_path.parent.mkdir(parents=True, exist_ok=True)
    with closing(sqlite3.connect(db_path)) as conn:
        ensure_schema(conn)
        if reset:
            reset_events(conn)

        rows_to_insert = []
        now = datetime.now(timezone.utc)

        for i in range(n_users):
            uid = uuid.uuid4().hex
            experiment = random.choice(["discount", "form_length"])
            variant = random.choice(["A", "B"])

            if experiment == "discount":
                p_convert = discount_rates[0 if variant == "A" else 1]
            else:
                p_convert = form_rates[0 if variant == "A" else 1]

            ts_imp = (now - timedelta(seconds=n_users - i)).isoformat(
                timespec="seconds"
            )

            rows_to_insert.append(
                (
                    uid,
                    experiment,
                    variant,
                    "impression",
                    ts_imp,
                    "Simulator/1.0",
                    "",
                    None,
                )
            )

            if random.random() < p_convert:
                ts_conv = (
                    now
                    - timedelta(seconds=n_users - i)
                    + timedelta(seconds=random.randint(5, 120))
                ).isoformat(timespec="seconds")
                rows_to_insert.append(
                    (
                        uid,
                        experiment,
                        variant,
                        "conversion",
                        ts_conv,
                        "Simulator/1.0",
                        "",
                        "simulated=1",
                    )
                )

        conn.executemany(
            """
            INSERT INTO events
                (user_id, experiment, variant, event_type,
                 created_at, user_agent, referrer, payload)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """,
            rows_to_insert,
        )
        conn.commit()

        total_events = conn.execute("SELECT COUNT(*) FROM events").fetchone()[0]

    print(
        f"Simulated {n_users} users → wrote {len(rows_to_insert):,} events "
        f"to {db_path} (total events now: {total_events:,})."
    )


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument(
        "--db",
        type=Path,
        default=Path(os.environ.get("AB_DB_PATH", str(DEFAULT_DB_PATH))),
        help="Path to the SQLite database (default: ab_test.db).",
    )
    p.add_argument("--n", type=int, default=2000, help="Number of users to simulate.")
    p.add_argument("--seed", type=int, default=42, help="Random seed.")
    p.add_argument(
        "--reset",
        action="store_true",
        help="Delete all existing events before simulating.",
    )
    p.add_argument("--discount-a", type=float, default=0.10)
    p.add_argument("--discount-b", type=float, default=0.15)
    p.add_argument("--form-a", type=float, default=0.50)
    p.add_argument("--form-b", type=float, default=0.35)
    args = p.parse_args(argv)

    simulate(
        args.db,
        args.n,
        discount_rates=(args.discount_a, args.discount_b),
        form_rates=(args.form_a, args.form_b),
        seed=args.seed,
        reset=args.reset,
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
